### PR TITLE
Add /user/{user_id}/shared_rooms/ api

### PR DESCRIPTION
**Pull request by Half-Shot**
This change adds a yet-to-be-MSCed unstable feature for fetching rooms you have in common with a user. Since synapse already implements this at the storage level, this API just exposes this as an endpoint in order to provide support for https://github.com/vector-im/riot-web/issues/14339.

[MSC 2666](https://github.com/matrix-org/matrix-doc/pull/2666)

#### Leftover things to implement from the MSC

- Reject `user_id` == `other_user_id`
- Replace `rooms` with `joined`.

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [x] Code style is correct (run the [linters](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#code-style))

### CODE REVIEW

**Comment by Half-Shot** [comment]
I seem to be having some CI challenges:
- isort has jumped to `isort==5.0.3`, wheras a PR from two days ago in this repo was using `isort==4.3.21`
- `black` wants `return 200, {"rooms": rooms,}` and `flake8` wants ` return 200, {"rooms": rooms, }`

File: synapse/rest/__init__.py [hunk 1]
```diff
@@ -113,6 +114,7 @@ def register_servlets(client_resource, hs):
         devices.register_servlets(hs, client_resource)
         thirdparty.register_servlets(hs, client_resource)
         sendtodevice.register_servlets(hs, client_resource)
+        shared_rooms.register_servlets(hs, client_resource)
```
Comment (anoadragon453) [inline synapse/rest/__init__.py:117]: Hm, I may be inclined to combine the `filter` and `shared_rooms` servlets here as they both start with `/user`, but that could possibly be deferred to a later PR.

File: synapse/rest/client/v2_alpha/shared_rooms.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+    def __init__(self, hs):
+        super(UserSharedRoomsServlet, self).__init__()
+        self.auth = hs.get_auth()
+        self.store = hs.get_datastore()
+
+    async def on_GET(self, request, user_id, other_user_id):
+        requester = await self.auth.get_user_by_req(request)
+        if user_id != requester.user.to_string():
+            raise AuthError(403, "Cannot get shared rooms for other users.")
```
Comment (anoadragon453) [inline synapse/rest/client/v2_alpha/shared_rooms.py:43]: I think this is wordly slightly unclearly. How about

```suggestion
            raise AuthError(403, "Cannot get shared rooms between other users.")
```

File: synapse/rest/client/v2_alpha/shared_rooms.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+logger = logging.getLogger(__name__)
+
+
+class UserSharedRoomsServlet(RestServlet):
+    """
+    GET /user/{user_id}/shared_rooms/{other_user_id} HTTP/1.1
+    """
+
+    PATTERNS = client_patterns(
+        "/user/(?P<user_id>[^/]*)/shared_rooms/(?P<other_user_id>[^/]*)",
```
Comment (anoadragon453) [inline synapse/rest/client/v2_alpha/shared_rooms.py:31]: This is more a question for the spec side, but why offer users the ability to choose the user_id that's put here if they're only allowed to put their own?

If it's for consistency's sake, I believe @ara4n has voiced that this was originally done for application services, but with AS tokens being a thing, there's no need to continue this behaviour in the future.

Comment (Half-Shot) [reply synapse/rest/client/v2_alpha/shared_rooms.py]: Yeah, it was for consistency. Let's also take it to the MSC because more :eyes:

File: synapse/rest/client/v2_alpha/shared_rooms.py (hunk 2 above)
Comment (Half-Shot) [reply synapse/rest/client/v2_alpha/shared_rooms.py:43]: I dunno, I think that sounds more confusing tbh. Probably need to #wordsmith it.

File: synapse/rest/__init__.py (hunk 1 above)
Comment (Half-Shot) [reply synapse/rest/__init__.py:117]: Ah, good point. I'm minded to defer it to keep this change contained though.

File: synapse/rest/client/v2_alpha/shared_rooms.py (hunk 3 above)
Comment (anoadragon453) [reply synapse/rest/client/v2_alpha/shared_rooms.py]: Lemme know when a link for an MSC exists then :P

File: synapse/rest/__init__.py (hunk 1 above)
Comment (anoadragon453) [reply synapse/rest/__init__.py:117]: Sounds good to me.

File: synapse/rest/client/v2_alpha/shared_rooms.py (hunk 3 above)
Comment (anoadragon453) [reply synapse/rest/client/v2_alpha/shared_rooms.py]: Ah, it does exist. Could you link it in the PR description please.

**Comment by richvdh** [comment]
@Half-Shot, @anoadragon453 : what's the state of play here?

**Comment by anoadragon453** [comment]
@richvdh https://github.com/matrix-org/matrix-doc/pull/2666 seems to have stalled on the spec core team's side. I'm not sure if this PR needs updating for the current state of the MSC. Can you confirm @Half-Shot?

The MSC has unstable prefixes clearly defined, so I think this PR can go ahead as long as it adheres to those.

**Comment by richvdh** [comment]
ok, will label it as awaiting changes until @Half-Shot gets a chance to update it.

**Comment by richvdh** [comment]
(mostly: please avoid the situation where both of you are waiting for the other to make the next move)

**Comment by Half-Shot** [comment]
Hi, sorry for taking a bit too long here. I believe we are stalled on the spec side, but I need to double check this PR conforms to the latest v of the spec. Will do so now.

**Comment by Half-Shot** [comment]
This should now conform to the spec. @anoadragon453 it should be good to go.

**Comment by anoadragon453** [comment]
@Half-Shot Thanks. Could you do a pull from develop to fix CI as well please?

File: changelog.d/7785.feature [hunk 4]
```diff
@@ -0,0 +1 @@
+Add `/user/{user_id}/shared_rooms/{other_user_id}` endpoint.
```
Comment (anoadragon453) [inline changelog.d/7785.feature:1]:
```suggestion
Add an endpoint to query your shared rooms with another user as an implementation of [MSC2666](https://github.com/matrix-org/matrix-doc/pull/2666).
```

File: synapse/rest/client/v2_alpha/shared_rooms.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        self.auth = hs.get_auth()
+        self.store = hs.get_datastore()
+
+    async def on_GET(self, request, user_id):
+        requester = await self.auth.get_user_by_req(request)
+        if user_id == requester.user.to_string():
+            raise SynapseError(
+                code=400,
+                msg="'user_id' must not be the authenticated user",
+                errcode=Codes.BAD_JSON,
```
Comment (anoadragon453) [inline synapse/rest/client/v2_alpha/shared_rooms.py]:
```suggestion
                errcode=Codes.INVALID_PARAM,
```

It would still be valid JSON.

File: synapse/rest/client/v2_alpha/shared_rooms.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    async def on_GET(self, request, user_id):
+        requester = await self.auth.get_user_by_req(request)
+        if user_id == requester.user.to_string():
+            raise SynapseError(
+                code=400,
+                msg="'user_id' must not be the authenticated user",
+                errcode=Codes.FORBIDDEN,
+            )
+
+        rooms = await self.store.get_rooms_in_common_for_users(
```
Comment (anoadragon453) [inline synapse/rest/client/v2_alpha/shared_rooms.py:49]: I'm afraid this method was removed in https://github.com/matrix-org/synapse/pull/8045, as it was unused ever since a [rewrite of the user directory](https://github.com/matrix-org/synapse/pull/4537) occurred to calculate the user directory iteratively instead of doing an expensive calculation per-request.

The [original implementation](https://github.com/matrix-org/synapse/pull/8045/files#diff-a91810f1149e0f4efc0f75a33cdcbefdL678-L701) for this method was rather expensive. From the new user directory code, I see that there are `users_who_share_private_rooms` and `users_in_public_rooms` tables, both of which include remote users. Methods using them can be found here: https://github.com/matrix-org/synapse/blob/f3fe6961b211d898aa347771df598c531fbca90c/synapse/storage/databases/main/user_directory.py#L638-L664

You'll need to add another that gets rooms in common between two users though. Might I also suggest that this method have a cache thrown on it? You can do so by throwing:

```
@cached()
```

on top of the method declaration.

Comment (Half-Shot) [reply synapse/rest/client/v2_alpha/shared_rooms.py]: I've reimplemented it as best as I can in pure SQL. It looks like it has legs, and a cache.

File: synapse/storage/databases/main/user_directory.py [hunk 7]
```diff
@@ -663,6 +663,42 @@ async def get_user_dir_rooms_user_is_in(self, user_id):
...
         return list(users)
 
+    @cached()
+    async def get_shared_rooms_for_users(self, user_id, other_user_id):
+        """
+        Returns the rooms that a user is in.
+
+        Args:
+            user_id(str): Must be a local user
+
```
Comment (anoadragon453) [inline synapse/storage/databases/main/user_directory.py:673]: `other_user_id`?

File: synapse/storage/databases/main/user_directory.py [hunk 8]
```diff
@@ -663,6 +663,42 @@ async def get_user_dir_rooms_user_is_in(self, user_id):
...
+        rows = await self.db_pool.execute(
+            "get_shared_rooms_for_users",
+            None,
+            SQL,
+            user_id,
+            other_user_id,
+            user_id,
+            other_user_id,
+        )
+        return list({row[0] for row in rows})
```
Comment (anoadragon453) [inline synapse/storage/databases/main/user_directory.py:700]: I assume you're doing a list->set->list to deduplicate results, but is it possible to get duplicates in this query?

File: synapse/rest/client/v2_alpha/shared_rooms.py [hunk 9]
```diff
@@ -43,10 +43,10 @@ async def on_GET(self, request, user_id):
             raise SynapseError(
                 code=400,
                 msg="'user_id' must not be the authenticated user",
```
Comment (anoadragon453) [inline synapse/rest/client/v2_alpha/shared_rooms.py:45]:
```suggestion
                msg="'user_id' must not be the authenticated user",
```

I still think this wording is a bit weird. Perhaps something friendlier like "You cannot get shared rooms with yourself"?

File: synapse/rest/client/v2_alpha/shared_rooms.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        self.store = hs.get_datastore()
+
+    async def on_GET(self, request, user_id):
+        requester = await self.auth.get_user_by_req(request)
+        if user_id == requester.user.to_string():
+            raise SynapseError(
+                code=400,
+                msg="'user_id' must not be the authenticated user",
+                errcode=Codes.FORBIDDEN,
+            )
```
Comment (anoadragon453) [inline synapse/rest/client/v2_alpha/shared_rooms.py:47]: You'll probably want to do some basic sanity checking here for `user_id`.

`UserID.from_string(user_id)` will do just that for you.

File: synapse/storage/databases/main/user_directory.py [hunk 11]
```diff
@@ -663,6 +663,42 @@ async def get_user_dir_rooms_user_is_in(self, user_id):
...
+
+        Args:
+            user_id(str): Must be a local user
+
+        Returns:
+            list: user_id
+        """
+        SQL = """
+            SELECT p1.room_id
+            FROM users_in_public_rooms as p1
```
Comment (anoadragon453) [inline synapse/storage/databases/main/user_directory.py:679]: I'm afraid there's still the slight niggle of this code relying on the user directory updating processes, in that we should tell users that if it's [disabled](https://github.com/matrix-org/synapse/blob/bd6dc17221741d4ceae05ae769a70696ae939336/synapse/handlers/user_directory.py#L95) then this will break as well.

This [config option](https://github.com/matrix-org/synapse/blob/5c43c43240a85ca6b65ad327b6a5b1c9e29bd653/synapse/config/server.py#L141-L143) controls that variable. It's not listed in the sample config file as it's really only used when you're splitting user directory handling out to a user_dir worker.

Needless to say, we should document this somewhere. The place to start would be to note that `^/_matrix/client/unstable/user/.*/shared_rooms/` is also [handled by the user directory worker](https://github.com/matrix-org/synapse/blob/develop/docs/workers.md#synapseappuser_dir).

**Review by anoadragon453 — CHANGES_REQUESTED** [review]
I'd also like to see a unit test for this endpoint as well. Making one isn't too hard though. See something like: https://github.com/matrix-org/synapse/blob/32e7c9e7f20b57dd081023ac42d6931a8da9b3a3/tests/rest/client/test_identity.py

I'd create two users, create one or more private rooms, one or more public rooms and see if you get the right responses.

Some common edge cases could be:

* Checking that a room ID doesn't appear after you leave a room.
* Checking that A and B, and B and C share rooms, but A and C correctly do not.

File: synapse/storage/databases/main/user_directory.py (hunk 8 above)
Comment (richvdh) [reply synapse/storage/databases/main/user_directory.py:700]: (also: why not just return a set?)

File: synapse/storage/databases/main/user_directory.py [hunk 12]
```diff
@@ -663,6 +663,42 @@ async def get_user_dir_rooms_user_is_in(self, user_id):
...
+                AND p1.user_id = ?
+                AND p2.user_id = ?
+            UNION
+            SELECT room_id
+            FROM users_who_share_private_rooms
+            WHERE
+                user_id = ?
+                AND other_user_id = ?;
+        """
+        rows = await self.db_pool.execute(
```
Comment (anoadragon453) [inline synapse/storage/databases/main/user_directory.py:691]: We tend to use a txn function paired with `runInteraction` to execute DB queries, rather than running `execute` on the `db_pool` itself. See other examples in this file and elsewhere.

**Review by anoadragon453 — CHANGES_REQUESTED** [review]

File: synapse/storage/databases/main/user_directory.py (hunk 8 above)
Comment (Half-Shot) [reply synapse/storage/databases/main/user_directory.py:700]: it's not possible to get a duplicate in this query because it's a union? Do I still need to return a set?

Comment (Half-Shot) [reply synapse/storage/databases/main/user_directory.py:700]: Well, I made it a set anyway.

File: tests/rest/client/v2_alpha/test_shared_rooms.py [hunk 13]
```diff
@@ -0,0 +1,143 @@
+# -*- coding: utf-8 -*-
+# Copyright 2018 New Vector
```
Comment (anoadragon453) [inline tests/rest/client/v2_alpha/test_shared_rooms.py:2]:
```suggestion
# Copyright 2020 Half-Shot
```

File: tests/rest/client/v2_alpha/test_shared_rooms.py [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+        config = self.default_config()
+        config["update_user_directory"] = True
+        return self.setup_test_homeserver(config=config)
+
+    def prepare(self, reactor, clock, hs):
+        self.store = hs.get_datastore()
+        self.handler = hs.get_user_directory_handler()
+
+    def _get_shared_rooms(self, token, other_user):
+        # Assert user directory is not empty
```
Comment (anoadragon453) [inline tests/rest/client/v2_alpha/test_shared_rooms.py:45]: Not sure if this comment belongs here?

File: tests/rest/client/v2_alpha/test_shared_rooms.py [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+        u1 = self.register_user("user1", "pass")
+        u1_token = self.login(u1, "pass")
+        u2 = self.register_user("user2", "pass")
+        u2_token = self.login(u2, "pass")
+
+        room = self.helper.create_room_as(u1, is_public=True, tok=u1_token)
+        self.helper.invite(room, src=u1, targ=u2, tok=u1_token)
+        self.helper.join(room, user=u2, tok=u2_token)
+
+        # Assert user directory is not empty
```
Comment (anoadragon453) [inline tests/rest/client/v2_alpha/test_shared_rooms.py:69]: Copy/paste error? And a few more below?

File: tests/rest/client/v2_alpha/test_shared_rooms.py [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+        The shared room list between two users should contain both public and private
+        rooms.
+        """
+        u1 = self.register_user("user1", "pass")
+        u1_token = self.login(u1, "pass")
+        u2 = self.register_user("user2", "pass")
+        u2_token = self.login(u2, "pass")
+
+        room_public = self.helper.create_room_as(u1, is_public=False, tok=u1_token)
+        room_private = self.helper.create_room_as(u2, is_public=True, tok=u2_token)
```
Comment (anoadragon453) [inline tests/rest/client/v2_alpha/test_shared_rooms.py:106]: The names of these two seem to be flipped.

File: tests/rest/client/v2_alpha/test_shared_rooms.py [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+
+    servlets = [
+        login.register_servlets,
+        synapse.rest.admin.register_servlets_for_client_rest_resource,
+        room.register_servlets,
+        shared_rooms.register_servlets,
+    ]
+
+    def make_homeserver(self, reactor, clock):
+
```
Comment (anoadragon453) [inline tests/rest/client/v2_alpha/test_shared_rooms.py:35]:
```suggestion
```

**Review by anoadragon453 — COMMENTED** [review]
Looking great, and thanks for writing the tests :)

Couple small things below, and we should add a note in [UPGRADE.rst](https://github.com/matrix-org/synapse/blob/develop/UPGRADE.rst) about the new endpoint being added to the user dir worker. I can do that for you.

File: synapse/rest/__init__.py [hunk 18]
```diff
@@ -127,3 +126,6 @@ def register_servlets(client_resource, hs):
         synapse.rest.admin.register_servlets_for_client_rest_resource(
             hs, client_resource
         )
+
+        # unstable
+        shared_rooms.register_servlets(hs, client_resource)
```
Comment (clokep) [inline synapse/rest/__init__.py:131]:
```suggestion
        shared_rooms.register_servlets(hs, client_resource)

```

File: synapse/storage/databases/main/user_directory.py [hunk 19]
```diff
@@ -664,6 +664,48 @@ async def get_user_dir_rooms_user_is_in(self, user_id):
...
+                (user_id, other_user_id, user_id, other_user_id),
+            )
+            rows = self.db_pool.cursor_to_dict(txn)
+            return rows
+
+        rows = await self.db_pool.runInteraction(
+            "get_shared_rooms_for_users", _get_shared_rooms_for_users_txn
+        )
+
+        return set(row["room_id"] for row in rows)
```
Comment (Half-Shot) [inline synapse/storage/databases/main/user_directory.py:707]: Confused? Can this be simplified?

Comment (tulir) [reply synapse/storage/databases/main/user_directory.py:707]:
```suggestion
        return {row["room_id"] for row in rows}
```

Comment (Half-Shot) [reply synapse/storage/databases/main/user_directory.py:707]: Aha! Thanks

**Comment by Half-Shot** [comment]
Well, not sure how I have broken that test 😔

File: synapse/rest/client/v2_alpha/shared_rooms.py [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+logger = logging.getLogger(__name__)
+
+
+class UserSharedRoomsServlet(RestServlet):
+    """
+    GET /uk.half-shot.msc2666/user/shared_rooms/{user_id} HTTP/1.1
+    """
+
+    PATTERNS = client_patterns(
+        "/uk.half-shot.msc2666/user/shared_rooms/(?P<user_id>[^/]*)",
```
Comment (anoadragon453) [inline synapse/rest/client/v2_alpha/shared_rooms.py:32]: Sorry, just noticed this and your MSC [don't match up](https://github.com/matrix-org/matrix-doc/blob/hs/shared-rooms/proposals/2666-get-rooms-in-common.md#unstable-prefix). You'll want to put `/unstable/` before `/uk.half-shot.msc2666`. This matches the unstable endpoint definitions laid out in [MSC2324](https://github.com/matrix-org/matrix-doc/blob/master/proposals/2324-when-to-ship.md).

You can do so by adding `unstable=True` to this function call (and at that point the `# This is an unstable feature` comment is likely redundant).

File: UPGRADE.rst [hunk 21]
```diff
@@ -1,3 +1,16 @@
+Upgrading to v1.20.0
+====================
+
+Shared rooms endpoint (MSC2666)
+------------------------------
```
Comment (anoadragon453) [inline UPGRADE.rst:5]:
```suggestion
-------------------------------
```

File: synapse/storage/databases/main/user_directory.py [hunk 22]
```diff
@@ -664,6 +664,48 @@ async def get_user_dir_rooms_user_is_in(self, user_id):
...
+                INNER JOIN users_in_public_rooms as p2
+                    ON p1.room_id = p2.room_id
+                    AND p1.user_id = ?
+                    AND p2.user_id = ?
+                UNION
+                SELECT room_id
+                FROM users_who_share_private_rooms
+                WHERE
+                    user_id = ?
+                    AND other_user_id = ?;
```
Comment (anoadragon453) [inline synapse/storage/databases/main/user_directory.py:696]:
```suggestion
                    AND other_user_id = ?
```

**Review by anoadragon453 — CHANGES_REQUESTED** [review]
Sorry, just [one small thing](https://www.youtube.com/watch?v=iMfoM9Y8viE) left. Otherwise lgtm!

File: synapse/rest/client/v2_alpha/shared_rooms.py (hunk 20 above)
Comment (Half-Shot) [reply synapse/rest/client/v2_alpha/shared_rooms.py:32]: I don't need to. `unstable` [is True by default](https://github.com/matrix-org/synapse/blob/53f7b49f5b09aca3900c520a142496da31fc1ee3/synapse/rest/client/v2_alpha/_base.py#L32). I don't think I need to stick /unstable/ before the path either.

File: UPGRADE.rst [hunk 23]
```diff
@@ -1,3 +1,16 @@
+Upgrading to v1.20.0
+====================
+
+Shared rooms endpoint (MSC2666)
+-------------------------------
+
+This release contains a new unstable endpoint `/_matrix/client/uk.half-shot.msc2666/user/shared_rooms/.*`
```
Comment (anoadragon453) [inline UPGRADE.rst:7]:
```suggestion
This release contains a new unstable endpoint `/_matrix/client/unstable/uk.half-shot.msc2666/user/shared_rooms/.*`
```

File: docs/workers.md [hunk 24]
```diff
@@ -380,6 +380,7 @@ Handles searches in the user directory. It can handle REST endpoints matching
 the following regular expressions:
 
     ^/_matrix/client/(api/v1|r0|unstable)/user_directory/search$
+    ^/_matrix/client/uk.half-shot.msc2666/user/shared_rooms/.*$
```
Comment (anoadragon453) [inline docs/workers.md:383]:
```suggestion
    ^/_matrix/client/unstable/uk.half-shot.msc2666/user/shared_rooms/.*$
```

**Comment by Half-Shot** [comment]
Ugh, this is getting so tedious.

**Review by anoadragon453 — APPROVED** [review]
Just these then, and then lgtm! :tada:

File: docs/workers.md (hunk 24 above)
Comment (tulir) [inline docs/workers.md:383]:
```suggestion
    ^/_matrix/client/unstable/uk.half-shot.msc2666/user/shared_rooms/.*$
```

File: synapse/rest/client/v2_alpha/shared_rooms.py (hunk 20 above)
Comment (anoadragon453) [reply synapse/rest/client/v2_alpha/shared_rooms.py:32]: Ah! My bad. Just a couple doc updates then and we're good to go!

File: docs/workers.md (hunk 24 above)
Comment (tulir) [reply docs/workers.md:383]: anoa was a second faster :(

Comment (Half-Shot) [reply docs/workers.md]: @tulir got there first :p

**Comment by anoadragon453** [comment]
> Ugh, this is getting so tedious.

I know it was a lot of work - but in the end we ended up with a more efficient SQL query, tests and documentation :slightly_smiling_face:

**Comment by Half-Shot** [comment]
M

> > Ugh, this is getting so tedious.
> 
> I know it was a lot of work - but in the end we ended up with a more efficient SQL query, tests and documentation slightly_smiling_face

oh I meant the CI failing. Turns out I forgot to merge. I think it's all good now :crossed_fingers:

**Comment by richvdh** [comment]
Note that https://github.com/matrix-org/synapse/pull/8290 removed this endpoint from `workers.md`
